src/Microsoft.ML.CpuMath/SseIntrinsics.cs

```diff
@@ -0,0 +1,536 @@
+            }
+
+            return Sse.ConvertToSingle(result);
+        }
+
+        internal static float DotSU(float[] src, float[] dst, int[] idx, int count)
+        {
+            Vector128<float> result = Sse.SetZeroVector128();
+
+            unsafe
+            {
+                fixed (float* psrc = src)
+                fixed (float* pdst = dst)
+                fixed (int* pidx = idx)
+                {
+                    float* pSrcCurrent = psrc;
+                    float* pDstCurrent = pdst;
+                    int* pIdxCurrent = pidx;
+                    int* pEnd = pidx + count;
+
+                    while (pIdxCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> srcVector = Load4(pSrcCurrent, pIdxCurrent);
+                        Vector128<float> dstVector = Sse.LoadVector128(pDstCurrent);
+
+                        result = Sse.Add(result, Sse.Multiply(srcVector, dstVector));
+
+                        pIdxCurrent += 4;
+                        pDstCurrent += 4;
+                    }
+
+                    if (Sse3.IsSupported)
+                    {
+                        // SSE3 is supported.
+                        result = Sse3.HorizontalAdd(result, result);
+                        result = Sse3.HorizontalAdd(result, result);
+                    }
+                    else
+                    {
+                        // SSE3 is not supported.
+                        result = Sse.Add(result, Sse.MoveHighToLow(result, result));
+                        result = Sse.Add(result, Sse.MoveHighToLow(result, Sse.UnpackLow(result, result)));
+                    }
+
+
+                    while (pIdxCurrent < pEnd)
+                    {
+                        Vector128<float> srcVector = Load1(pSrcCurrent, pIdxCurrent);
+                        Vector128<float> dstVector = Sse.LoadScalarVector128(pDstCurrent);
+
+                        result = Sse.AddScalar(result, Sse.MultiplyScalar(srcVector, dstVector));
+
+                        pIdxCurrent++;
+                        pDstCurrent++;
+                    }
+                }
+            }
+
+            return Sse.ConvertToSingle(result);
+        }
+
+        internal static float Dist2(float[] src, float[] dst, int count)
+        {
+            Vector128<float> SqDistanceVector = Sse.SetZeroVector128();
+
+            unsafe
+            {
+                fixed (float* psrc = src)
+                fixed (float* pdst = dst)
+                {
+                    float* pSrcCurrent = psrc;
+                    float* pDstCurrent = pdst;
+                    float* pEnd = psrc + count;
+
+                    while (pSrcCurrent + 4 <= pEnd)
+                    {
+                        Vector128<float> distanceVector = Sse.Subtract(Sse.LoadVector128(pSrcCurrent), Sse.LoadVector128(pDstCurrent));
+                        SqDistanceVector = Sse.Add(SqDistanceVector, Sse.Multiply(distanceVector, distanceVector));
```

Comment:
I think there might have been other cases above as well, by why aren't you just doing `SqDistanceVector = Sse.Multiply(distanceVector, distanceVector)`